File: Algorithms/Dynamic-Programming/Largest_Sum_Contiguous_Subarray/LSCS.cpp
#include<iostream>

int Kadane(int arr[], int size)
{
  int max_so_far=0;
  int max_ending_here=0;

  for(int i=0; i<size; i++)
  {
    max_ending_here= max_ending_here + arr[i];
    if(max_so_far<max_ending_here)
      max_so_far= max_ending_here;
    if(max_ending_here<0)
      max_ending_here=0;
  }

  return max_so_far;
}

File: Algorithms/Dynamic-Programming/Largest_Sum_Contiguous_Subarray/LSCS.js
  let allPositives = arr => arr.every(n => n > 0)
  let allNegatives = arr => arr.every(n => n < 0)
  let sum = arr => arr.reduce((curr_max, max_so_far) => curr_max + max_so_far, 0)
  
  var maxSequence = function(arr){
    if(arr.length === 0 || allNegatives(arr)) return 0;
    if(allPositives(arr)) return sum(arr);
  
    var curr_max = 0, max_so_far = 0;
  
    for(var i = 0; i < arr.length; i++){  
      curr_max = Math.max(0, curr_max + arr[i]);
      max_so_far = Math.max(curr_max, max_so_far);
    }
    return max_so_far;
  }